README.md
# BD-Traffic
Descargar base de datos en el link compartido

#Pendientes
##Clasificación de Timestamps  [X]
Al clasificar por segmentos, aparecen todos los timestamps del día para cada bus. Sin embargo, se deben separar en grupos que competen ese momento en el tiempo.
encontrar la manera de clasificar o separar timestamps en bus_times.py
ideas: MLib? granularizacion? bucketing?

solución: 
  * [x] ordenar valores de menor a mayor. (probar algoritmo TimSort)
  * [x] hacer cadenas o listas enlazadas con valores sucesivos con la condición que entre cada valor sólo haya 40 segundos de diferencia.
  * [x] devolver sólo el par de timestamp mediano y el tiempo de recorrido (max-min).
  * [x] mapear cada valor de lista con sus llaves iniciales para tener la tupla {BusID,"segmento","periodo",TiempoRecorrido}
  * [x] la tupla deberá ser genérica para poder moldear los resultados según petición

##Generar csv/RDD para demo [ ]
Se ha generado un CSV con los valores basicos sin modificar
solución:
  * [x] generar una tupla con buckets de cada media hora.
  
##Generar tablas de segmentos/paraderos por ruta(intentar encontrar orden de paraderos) [ ]
##Generación de Periodos según granularización temporal (media hora, hora, dia, año) [ ]
##Generación de Periodos según granularización física (segmento, ruta) [ ]


history.txt
   23  clear
   24  ls
   25  clear
   26  cd
   27  clear
   28  clear
   29  ls
   30  clear
   31  ls
   32  cd noa/
   33  ls
   34  cd so
   35  cd spark-gpu/
   36  ls
   37  cd ..
   38  ls
   39  clear
   40  ls
   41  cd Documents/
   42  ls
   43  clear
   44  ls
   45  clear
   46  ls
   47  cd
   48  ls
   49  clear
   50  ls
   51  su mpiuser 
   52  clear
   53  ls
   54  clear
   55  cssh clusterChess
   56  clear
   57  cssh clusterChess
   58  exit
   59  clear
   60  ls
   61  clear
   62  cd /mnt/mpiuser/
   63  ls
   64  clear
   65  cd
   66  cssh clusterChess
   67  cd scripts/
   68  ls
   69  vim poweroff.sh 
   70  ./poweroff.sh 
   71  cleear
   72  clear
   73  ls
   74  cd /mnt/mpiuser/
   75  ls
   76  cd runsRenzo/
   77  ls
   78  pwd
   79  ping Queen
   80  ping Pawn01 
   81  clear
   82  ping Pawn01 
   83  clear
   84  cd
   85  cd scripts/
   86  ls
   87  vim poweroff.sh 
   88  ping www.google.com
   89  vim poweroff.sh 
   90  clear
   91  ls
   92  ./poweroff.sh 
   93  clear
   94  sudo poweroff
   95  ping www.google.com
   96  sudo poweroff
   97  clear
   98  exit
   99  clear
  100  exit
  101  clear
  102  ping www.google.com
  103  clear
  104  ls
  105  clear
  106  ping 8.8.8.8
  107  clear
  108  ping www.google.com
  109  clear
  110  exit
  111  clear
  112  ssh Queen
  113  exit
  114  clear
  115  ls
  116  clear
  117  ls
  118  clear
  119  cd scripts/
  120  ls
  121  vim conectados.py 
  122  python conectados.py 
  123  clear
  124  cd
  125  cd /mnt/mpiuser/
  126  ls
  127  cd runsRenzo/
  128  ls
  129  cd ..
  130  ls
  131  clear
  132  ls
  133  cd phi-GPU-v3.0/
  134  ls
  135  clear
  136  ls
  137  clear
  138  ls
  139  vim bh-param.plt 
  140  ls
  141  clear
  142  ls
  143  cd 4th-CPU/
  144  ls
  145  cd 1nodo/
  146  ls
  147  cd 128k/
  148  ls
  149  clear
  150  ls
  151  vim log.4th.128kdat.1n.1st 
  152  ls
  153  cat log.4th.128kdat.1n.1st 
  154  cat log.4th.128kdat.1n.2nd 
  155  cat log.4th.128kdat.1n.3rd 
  156  cat log.4th.128kdat.1n.4th 
  157  cat log.4th.128kdat.1n.5th 
  158  ls
  159  cd ..
  160  ls
  161  cd ..
  162  ls
  163  cd 2nodo/
  164  ls
  165  cd 128k/
  166  ls
  167  clear
  168  ls
  169  cd ..
  170  ls
  171  cd ..
  172  ls
  173  cd 4nodo/
  174  ls
  175  cd 128k/
  176  ls
  177  cat log.4th.128kdat.4n.5th 
  178  cat log.4th.128kdat.4n.3rd 
  179  cat log.4th.128kdat.4n.1st 
  180  cat log.4th.128kdat.4n.2nd 
  181  cd ..
  182  ls
  183  cd 256k/
  184  ls
  185  cd ..
  186  ls
  187  cd 8k/
  188  ls
  189  cat log.4th.8kdat.4n.5th 
  190  cat log.4th.8kdat.4n.2nd 
  191  cat log.4th.8kdat.4n.3rd 
  192  ls
  193  cd ..
  194  ls
  195  cd ..
  196  ls
  197  cd 8nodo/
  198  ls
  199  cd 128k/
  200  ls
  201  cat log.4th.128kdat.8n.1st 
  202  cat log.4th.128kdat.8n.2nd 
  203  cat log.4th.128kdat.8n.3rd 
  204  cd ..
  205  ls
  206  cd ..
  207  ls
  208  cd ..
  209  ls
  210  vim test128k-CPU.sh 
  211  vim cpu-4th 
  212  vim vim test128k-CPU.sh 
  213  vim test128k-CPU.sh 
  214  ls *.cfg
  215  vim phi-GPU6.cfg 
  216  vim data.inp 
  217  vim phi-GPU6.cfg 
  218  vim phi-GPU4.cfg 
  219  vim test8k-CPU.sh 
  220  ls
  221  vim machines1 
  222  vim machines2 
  223  vim machines3
  224  vim machines4
  225  vim machines6
  226  vim machines8 
  227  clear
  228  ls
  229  clear
  230  ls
  231  vim test128k-CPU.sh 
  232  vim phi-GPU.cpp 
  233  clear
  234  ls
  235  vim machines1 
  236  vim machines2 
  237  vim machines3
  238  vim machines4
  239  vim machines8
  240  clear
  241  ls
  242  ls *.cfg
  243  vim phi-GPU4.cfg 
  244  vim phi-GPU6.cfg 
  245  vim phi-GPU8.cfg 
  246  vim data.inp 
  247  clear
  248  htop
  249  clear
  250  ls
  251  htop
  252  clear
  253  htop
  254  clear
  255  htop
  256  ls
  257  clear
  258  cd
  259  cd scripts/
  260  ls
  261  python conectados.py 
  262  clear
  263  htop
  264  python conectados.py 
  265  htop
  266  cd /mnt/mpiuser/
  267  ls
  268  cd phi-GPU-v3.0/
  269  ls
  270  cd 4th-CPU/
  271  ls
  272  cd 1nodo/
  273  ls
  274  cd 128k/
  275  ls
  276  vim log.4th.128kdat.1n.3rd 
  277  vim log.4th.128kdat.1n.4th 
  278  cd ..
  279  ls
  280  vim 256k/
  281  cd 256k/
  282  ls
  283  cd ..
  284  ls
  285  cd ..
  286  ls
  287  cd 8nodo/
  288  ls
  289  cd 128k/
  290  ls
  291  vim log.4th.128kdat.8n.1st 
  292  ls
  293  clear
  294  ls
  295  vim log.4th.128kdat.8n.5th 
  296  vim log.4th.128kdat.8n.4th 
  297  vim log.4th.128kdat.8n.5th 
  298  vim log.4th.128kdat.8n.3rd 
  299  vim log.4th.128kdat.8n.2nd 
  300  pwd
  301  cd ..
  302  ls
  303  cd ..
  304  ls
  305  cd 2nodo/
  306  ls
  307  cd 128k/
  308  ls
  309  vim log.4th.128kdat.2n.1st 
  310  vim log.4th.128kdat.2n.2nd 
  311  vim log.4th.128kdat.2n.3rd 
  312  vim log.4th.128kdat.2n.4th 
  313  cd ..
  314  ls
  315  cd ..
  316  ls
  317  cd 1nodo/
  318  ls
  319  cd 128k/
  320  ls
  321  vim log.4th.128kdat.1n.2nd 
  322  vim log.4th.128kdat.1n.3rd 
  323  cd ..
  324  cd ../2nodo/128k/
  325  vim log.4th.128kdat.2n.1st 
  326  vim log.4th.128kdat.2n.2nd 
  327  htop
  328  clear
  329  htop
  330  ls
  331  vim log.4th.128kdat.2n.1st 
  332  cd .. 
  333  cd ..
  334  cd 1nodo/
  335  ls
  336  cd 128k/
  337  ls
  338  vim log.4th.128kdat.1n.1st 
  339  vim log.4th.128kdat.1n.2nd 
  340  vim log.4th.128kdat.1n.3rd 
  341  ls
  342  cd ..
  343  cd ../2nodo/128k/
  344  ls
  345  vim log.4th.128kdat.2n.1st 
  346  clear
  347  htop
  348  clear
  349  ls
  350  vim log.4th.128kdat.2n.2nd 
  351  clear
  352  ssh Pawn01
  353  clear
  354  ls
  355  vim log.4th.128kdat.2n.1st 
  356  ls
  357  clear
  358  ls
  359  clear
  360  ls
  361  vim log.4th.128kdat.2n.1st 
  362  vim log.4th.128kdat.2n.2nd 
  363  clear
  364  vim log.4th.128kdat.2n.2nd 
  365  clear
  366  vim log.4th.128kdat.2n.2nd 
  367  clear
  368  htop
  369  clear
  370  ls
  371  vim log.4th.128kdat.2n.2nd 
  372  clear
  373  ls
  374  cd ..
  375  ls

MTAfilter.py
from pyspark import SparkContext, SparkConf
from pyspark.sql import SQLContext, Row
from datetime import datetime
import time

#Cuenta las comillas en la linea, si no se cumplen retorna falso

def countTab(x):
	c=0
	for i in x:
		for j in i:
			if (j=='\t'):
				c=c+1
	return c

#limpia las comillas
def cleanEl(x):
	r=''
	for i in x:
		for j in i:
			if (j!='"'):
				r=r+j
	return r	


#La Funcion formateara la fecha y hora a UNX

def forTim(x):
	fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
	x[2]=time.mktime(fecha.timetuple())	
	
	return x

#La funcion transformara los timestamps en bucket para separarlos por hora
def toHourBucket(x):
	horaEnSeg=60*60
	timeStamp=x[4]
	x[4]= timeStamp//horaEnSeg
	return x


###########################################################

#Inicio de Contexto Spark
conf = SparkConf().setMaster("spark://King:7077").setAppName("MTAFilter")

sc = SparkContext(conf = conf)

rdd = sc.textFile("hdfs://King:9000/user/bdata/mta_data/MTA-Bus-Time_.2014-08-01.txt")


#limpia ingresos repetidos
cleaned=rdd.distinct()


#descarta lineas corruptas que son aquellas que no cumplen las 26 comillas
discardCorrupted= cleaned.filter(lambda x: countTab(x)==10)


#va a mapear cada ingreso y le quitar las comillas
#myFile1 = discardCorrupted.map( lambda x: cleanEl(x))


#va a mapear cada ingreso y lo  partira en elementos a traves de la tabulacin
filteredT= discardCorrupted.map(lambda x:x.split('\t'))


#descarta las lineas que son headers
discardHeaders=filteredT.filter(lambda x : x[0]!="latitude" )

#formatea los tiempos a segundos en UNX
formatTime= discardHeaders.map(lambda x: forTim(x))

#solo mostrara los ingresos para una ruta
#getResultsFromOneRoute= formatTime.filter(lambda x: x[7]=='MTA NYCT_B63' )

#solo mostrara los ingresos para un bus
getResultsFromOneBus= formatTime.filter(lambda x: x[3]=='469')


#Comienzo de Tranformaciones para la Vista precomputada
#Formateo de Timestamp a Bucket de Hora
#HourBucket= NormalizeSpeedUndTime.map(lambda x: toHourBucket(x))


# mapear los resultados en llaves con dos valores para agruparlos luego. cada RDD tendra ((ID,BUCKET),SPEED)
#createKeys=HourBucket.map(lambda x: ((x[0],'h',x[4]),x[1]))


#Se agrupan los valores con todas las tuplas que le corresponden
#groupedKeys=createKeys.groupByKey()

#Se obtienen los valores maximos por hora
#maxPerHour= groupedKeys.mapValues(lambda x: max(x))

#Segunda parte Emision de granularidades para valores maximo por hora

#se genera un nuevo RDD con las granularidades por dia y el maximo por dia
#maxPerDay= maxPerHour.map(lambda (x,y): ((x[0],'d',x[2]//24),y)).groupByKey().mapValues(lambda x: max(x))

#se genera un nuevo RDD con las granularidades por semana y el maximo por semana
#maxPerWeek= maxPerDay.map(lambda (x,y): ((x[0],'s',x[2]//7),y)).groupByKey().mapValues(lambda x: max(x))

#se genera un nuevo RDD con las granularidades por mes y el maximo por mes

#REVISAR
#maxPerMonth= maxPerWeek.map(lambda (x,y): ((x[0],'m',x[2]//28),y)).groupByKey().mapValues(lambda x: max(x))


#allMaxRDD=maxPerHour.union(maxPerDay).union(maxPerWeek).union(maxPerMonth)

#allMaxRDD.saveAsTextFile("hdfs://King:9000/user/bdata/MHSPB_MasterDataset.txt")

#print("Emision de granularidades segun hora dia sem
for i in getResultsFromOneBus.take(100):
#for i in getResultsFromOneRoute.take(100):
        print(i)



batch_main.py

# coding: utf-8

# In[ ]:

from pyspark import SparkContext, SparkConf
from pyspark.sql import SQLContext, Row
import bus_times


# In[ ]:

def toCSV(x):
    y=str(x[0])
    for i in range(1,len(x)):
       y= y+","+str(x[i]) 
    return y
    
##########################################################################
#conf = SparkConf().setMaster("spark://King:7077").setAppName("batch_Main")
conf = SparkConf().setAppName("batch_Main")


sc = SparkContext(conf = conf)
sc.addPyFile("/home/bdata/sparkPrograms/BD-Traffic/bus_times.py")

#rdd = sc.textFile('file:///home/cloudera/Downloads/MTA-Bus-Time_.2014-08-01.txt')

#rdd = sc.textFile('hdfs://King:9000/user/bdata/mta_data/MTA-Bus-Time_.2014-08-01.txt')
rdd = sc.textFile('hdfs://King:9000/user/bdata/mta_data/MTA-Bus-Time_.2014-08-*')

classTuple= bus_times.mainFilter(rdd)
halfHourBucket=classTuple.map(lambda x: bus_times.toHalfHourBucket(list(x)))

##########################################################################   
#CSV
#classTuple.map(lambda x: toCSV(x)).saveAsTextFile('hdfs:/user/bdata/demo_test.csv')

halfHourBucket.map(lambda x: toCSV(x)).saveAsTextFile('hdfs://King:9000/user/bdata/demo_test_august.csv')


###########################################################################


###########################################################################
#RDD
#Se genera una tupla para clasificacion con los valores hora, diaSemana, busID,orientation, nextStop, Route
#classTuple= formatTime.map(lambda x: (x[2],x[11],x[3],x[5],x[10],x[7]))

#classTuple.saveAsTextFile('hdfs:/user/bdata/classification_buses_test.txt')


#Para correr pruebas quitar comentario
#for i in halfHourBucket.take(20):

#for i in getResultsFromOneBus.take(3):
#for i in getResultsFromOneRoute.take(100):
     
	# print(i)


# In[ ]:

#print(classTuple.count())


# In[ ]:




# In[ ]:





MTA_filter_main.py
from pyspark import SparkContext, SparkConf
from pyspark.sql import SQLContext, Row
from datetime import datetime
import time


# In[2]:

def countTab(x):
        c=0
        for i in x:
                for j in i:
                        if (j=='\t'):
                                c=c+1
        return c


def cleanEl(x):
        r=''
        for i in x:
                for j in i:
                        if (j!='"'):
                                r=r+j
        return r




def forTim(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=time.mktime(fecha.timetuple())

#probar
def hours(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=fecha.hour
	x.append(fecha.weekday())
        return x


def toHourBucket(x):
        horaEnSeg=60*60
        timeStamp=x[4]
        x[4]= timeStamp//horaEnSeg
        return x



def mainFilter(rdd):
	#limpia ingresos repetidos
	cleaned=rdd.distinct()


	#descarta lineas corruptas que son aquellas que no cumplen las 26 comillas
	discardCorrupted= cleaned.filter(lambda x: countTab(x)==10)

	#va a mapear cada ingreso y lo  partira en elementos a traves de la tabulacin
	filteredT= discardCorrupted.map(lambda x:x.split('\t'))

	#descarta las lineas que son headers
	discardHeaders=filteredT.filter(lambda x : x[0]!="latitude" )

	#formatea los tiempos a segundos en UNX
	#formatTime= discardHeaders.map(lambda x: forTim(x))
	
	#formatea los tiempos a horas 
	formatTime= discardHeaders.map(lambda x: hours(x))


	#Se genera una tupla para clasificacion con los valores hora, diaSemana, busID,orientation, nextStop, Route
	classTuple= formatTime.map(lambda x: (x[2],x[11],x[3],x[5],x[10],x[7]))

	return classTuple


##########################################################################
#conf = SparkConf().setMaster("spark://King:7077").setAppName("MTAMain")

#sc = SparkContext(conf = conf)

#rdd = sc.textFile('hdfs://King:9000/user/bdata/mta_data/MTA-Bus-Time_.2014-08-01.txt')


###########################################################################






















#collectedTuple=classTuple.collect()

#classTuple.saveAsTextFile('hdfs:/user/cloudera/wordcount/classification_buses.csv')
#classTuple.saveAsTextFile('file:///home/cloudera/Downloads/classification_buses.csv')
#collectedTuple.saveAsTextFile('file:///home/cloudera/Downloads/classification_buses.csv')


# In[15]:

#for i in classTuple.take(3):
#for i in getResultsFromOneBus.take(3):
#for i in getResultsFromOneRoute.take(100):
#        print(i)


# In[ ]:





main_filter_test_vm.py

# coding: utf-8

# In[6]:

from pyspark import SparkContext, SparkConf
from pyspark.sql import SQLContext, Row
from datetime import datetime
import time


# In[2]:

def countTab(x):
        c=0
        for i in x:
                for j in i:
                        if (j=='\t'):
                                c=c+1
        return c


def cleanEl(x):
        r=''
        for i in x:
                for j in i:
                        if (j!='"'):
                                r=r+j
        return r




def forTim(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=time.mktime(fecha.timetuple())

#probar
def hours(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=fecha.hour
        x.append(fecha.weekday())
        return x


def toHourBucket(x):
        horaEnSeg=60*60
        timeStamp=x[4]
        x[4]= timeStamp//horaEnSeg
        return x



def mainFilter(rdd):
	#limpia ingresos repetidos
    cleaned=rdd.distinct()


	#descarta lineas corruptas que son aquellas que no cumplen las 26 comillas
    discardCorrupted= cleaned.filter(lambda x: countTab(x)==10)

	#va a mapear cada ingreso y lo  partira en elementos a traves de la tabulacin
    filteredT= discardCorrupted.map(lambda x:x.split('\t'))

	#descarta las lineas que son headers
    discardHeaders=filteredT.filter(lambda x : x[0]!="latitude" )

	#formatea los tiempos a segundos en UNX
	#formatTime= discardHeaders.map(lambda x: forTim(x))
	
	#formatea los tiempos a horas 
    formatTime= discardHeaders.map(lambda x: hours(x))

    #RDD
	#Se genera una tupla para clasificacion con los valores hora, diaSemana, busID,orientation, nextStop, Route
	#classTuple= formatTime.map(lambda x: (x[2],x[11],x[3],x[5],x[10],x[7]))
    
    #CSV
    #Se genera una tupla para clasificacion con los valores hora, diaSemana, busID,orientation, nextStop, Route
    classTuple= formatTime.map(lambda x: str(x[2])+","+str(x[11])+","+x[3]+","+x[5]+","+x[10]+","+x[7])

    return classTuple


##########################################################################
#conf = SparkConf().setMaster("spark://King:7077").setAppName("MTAMain")

#sc = SparkContext(conf = conf)

rdd = sc.textFile('file:///home/cloudera/Downloads/MTA-Bus-Time_.2014-08-01.txt')

classTuple=mainFilter(rdd)


#Descomentar la siguiente linea para guardar en hdfs como csv
classTuple.saveAsTextFile('hdfs:/user/cloudera/classification_buses.csv')
###########################################################################

#collectedTuple=classTuple.collect()

#classTuple.saveAsTextFile('hdfs:/user/bdata/classification_buses_test.txt')
#collectedTuple.saveAsTextFile('file:///home/cloudera/Downloads/classification_buses.csv')


#for i in classTuple.take(1000):
#for i in getResultsFromOneBus.take(3):
#for i in getResultsFromOneRoute.take(100):
#        print(i)


# In[ ]:


# In[ ]:





MTA_filter_main_clusterTest.py
from pyspark import SparkContext, SparkConf
from pyspark.sql import SQLContext, Row
from datetime import datetime
import time


# In[2]:

def countTab(x):
        c=0
        for i in x:
                for j in i:
                        if (j=='\t'):
                                c=c+1
        return c


def cleanEl(x):
        r=''
        for i in x:
                for j in i:
                        if (j!='"'):
                                r=r+j
        return r




def forTim(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=time.mktime(fecha.timetuple())

#probar
def hours(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=fecha.hour
	x.append(fecha.weekday())
        return x


def toHourBucket(x):
        horaEnSeg=60*60
        timeStamp=x[4]
        x[4]= timeStamp//horaEnSeg
        return x



def mainFilter(rdd):
	#limpia ingresos repetidos
	cleaned=rdd.distinct()


	#descarta lineas corruptas que son aquellas que no cumplen las 26 comillas
	discardCorrupted= cleaned.filter(lambda x: countTab(x)==10)

	#va a mapear cada ingreso y lo  partira en elementos a traves de la tabulacin
	filteredT= discardCorrupted.map(lambda x:x.split('\t'))

	#descarta las lineas que son headers
	discardHeaders=filteredT.filter(lambda x : x[0]!="latitude" )

	#formatea los tiempos a segundos en UNX
	#formatTime= discardHeaders.map(lambda x: forTim(x))
	
	#formatea los tiempos a horas 
	formatTime= discardHeaders.map(lambda x: hours(x))


	#Se genera una tupla para clasificacion con los valores hora, diaSemana, busID,orientation, nextStop, Route
	classTuple= formatTime.map(lambda x: (x[2],x[11],x[3],x[5],x[10],x[7]))

	return classTuple


##########################################################################
conf = SparkConf().setMaster("spark://King:7077").setAppName("MTAMain")

sc = SparkContext(conf = conf)

rdd = sc.textFile('hdfs://King:9000/user/bdata/mta_data/MTA-Bus-Time_.2014-08-01.txt')

classTuple=mainFilter(rdd)
###########################################################################



#collectedTuple=classTuple.collect()

#classTuple.saveAsTextFile('hdfs:/user/bdata/classification_buses_test.txt')
#classTuple.saveAsTextFile('file:///home/cloudera/Downloads/classification_buses.csv')
#collectedTuple.saveAsTextFile('file:///home/cloudera/Downloads/classification_buses.csv')


for i in classTuple.take(200):
#for i in getResultsFromOneBus.take(3):
#for i in getResultsFromOneRoute.take(100):
        print(i)


# In[ ]:





bus_times.py

from pyspark import SparkContext, SparkConf
from pyspark.sql import SQLContext, Row
from datetime import datetime
import time


# In[2]:

def countTab(x):
        c=0
        for i in x:
                for j in i:
                        if (j=='\t'):
                                c=c+1
        return c


def cleanEl(x):
        r=''
        for i in x:
                for j in i:
                        if (j!='"'):
                                r=r+j
        return r




def forTim(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=time.mktime(fecha.timetuple())
        return x

#probar
def hours(x):
        fecha=datetime.strptime(x[2], '%Y-%m-%d %H:%M:%S')
        x[2]=fecha.hour
        x.append(fecha.weekday())
        return x


def toHourBucket(x):
        horaEnSeg=60*60
        timeStamp=x[4]
        x[4]= timeStamp//horaEnSeg
        return x
    
    
def toHalfHourBucket(x):
        hora= (x[4] // (60*60))%24
        minuto= (x[4] //60)%60
        if (minuto < 30):
            minuto=0.0
        else:
            minuto=0.5 
        valor= hora+minuto
        x[4]=valor
        return x
    
# La funcion ordenara los Timestamps y debera agruparlos por periodos similares,
# los periodos son conformados por TS en orden ascendente cuya diferencia entre valores contiguos
# no sea mayor a 120 segundos

# los periodos que solo tienen 1 TS figuran con 0 (verificar si se mantiene de esta manera) 

def groupByPeriod(x):
    x=sorted(x)
    tiemposRecorrido=[]
    
    tsMinimo=x[0]
    tsMaximo=x[0]
    cMediana=1
    periodoTs=[x[0]]
    for i in range(0,len(x)-1):
                 
            
        if (x[i+1]-x[i] <= 120):
            
            tsMaximo=x[i+1]
            periodoTs.append(x[i+1])
            cMediana+=1
            
            if (i == len(x)-2):
                
                if (cMediana%2==0):
                    medianaTs=(periodoTs[(cMediana//2)-1] + periodoTs[cMediana//2])//2
                else:
                    medianaTs= periodoTs[cMediana//2]
                    
                tiemposRecorrido.append((medianaTs, tsMaximo-tsMinimo))
            
                
        else:
                
                
            if (cMediana%2==0):
                medianaTs=(periodoTs[(cMediana//2)-1] + periodoTs[cMediana//2])//2
            else:
                medianaTs= periodoTs[cMediana//2]
                
            tiemposRecorrido.append((medianaTs, tsMaximo-tsMinimo))
            tsMinimo=x[i+1]
            tsMaximo=x[i+1]
            cMediana=1
            periodoTs=[x[i+1]]
            
            if (i == len(x)-2):
                tiemposRecorrido.append((tsMinimo, 0))
    
    #return x    
    return tiemposRecorrido



def mainFilter(rdd):
	#limpia ingresos repetidos
    cleaned=rdd.distinct()


	#descarta lineas corruptas que son aquellas que no cumplen las 26 comillas
    discardCorrupted= cleaned.filter(lambda x: countTab(x)==10)

	#va a mapear cada ingreso y lo  partira en elementos a traves de la tabulacin
    filteredT= discardCorrupted.map(lambda x:x.split('\t'))

	#descarta las lineas que son headers
    discardHeaders=filteredT.filter(lambda x : x[0]!="latitude" )

	#formatea los tiempos a segundos en UNX
    formatTime= discardHeaders.map(lambda x: forTim(x))
	
	#formatea los tiempos a horas 
    #formatTime= discardHeaders.map(lambda x: hours(x))
    
    #devolver solo de un bus
    #getResultsFromOneBus= formatTime.filter(lambda x: x[3]=='4215')
    
    #getResultsFromOneRoute= getResultsFromOneBus.filter(lambda x: x[7]=='MTA NYCT_Q20B' )
    
    #getResultsFromOneStop= getResultsFromOneRoute.filter(lambda x: x[10] =='MTA_505010') 
    
    #RDD con llaves
    keyedTuple= formatTime.map(lambda x: ((x[3],(x[7],x[10],x[5])),x[2]))
    
    groupedKeys=keyedTuple.groupByKey().mapValues(lambda x: groupByPeriod(x))
    
    
    ##Armar tupla final utilizando indices de arrays en python
    flattenedGroups= groupedKeys.flatMapValues(lambda x: x).map(lambda x: (x[0][0],x[0][1][0],x[0][1][1],x[0][1][2],x[1][0],x[1][1]))
            
    #lambda x: (max(x),min(x),max(x)-min(x))
    #RDD
	#Se genera una tupla para clasificacion con los valores hora, diaSemana, busID,orientation, nextStop, Route
	#classTuple= formatTime.map(lambda x: (x[2],x[11],x[3],x[5],x[10],x[7]))
    
    #CSV
    #Se genera una tupla para clasificacion con los valores hora, diaSemana, busID,orientation, nextStop, Route
    #classTuple= formatTime.map(lambda x: str(x[2])+","+str(x[11])+","+x[3]+","+x[5]+","+x[10]+","+x[7])

    return flattenedGroups






